lint/base_linter/node_linter.py

```diff
@@ -146,11 +147,45 @@ def find_local_executable(self, start_dir, npm_name):
                     self.notify_failure()
                     raise linter.PermanentError()
 
-                # A 'package.json' not yet installed?
+                # No 'node_modules' folder ?
                 if not os.path.exists(os.path.join(path, 'node_modules', '.bin')):
                     is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
                     is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
                     if is_dep or is_dev_dep:
+
+                        # Perhaps this is a Yarn PnP project?
+                        yarn_path = shutil.which('yarn')
+                        yarn_lock_exists = os.path.exists(os.path.join(path, 'yarn.lock'))
+                        package_lock_exists = os.path.exists(os.path.join(path, 'package-lock.json'))
+                        pnp_js_exists = os.path.exists(os.path.join(path, '.pnp.js'))
+                        pnp_in_manifest = bool(manifest.get('installConfig', {}).get('pnp'))
+
+                        if yarn_path is None:
+                            if yarn_lock_exists and (pnp_in_manifest or pnp_js_exists):
+                                logger.warning(
+                                    "This seems like a Yarn PnP project. However, finding "
+                                    "a Yarn executable failed. Make sure to install Yarn first."
+                                )
+                                self.notify_failure()
+                                raise linter.PermanentError()
+                        else:
+                            if yarn_lock_exists and pnp_js_exists:
+                                self.context['project_root'] = path
+                                return [yarn_path, 'run', '--silent', npm_name]
+                            elif yarn_lock_exists and not package_lock_exists:
+                                logger.warning(
+                                    "Skipping '{}' for now which is listed as a {} "
+                                    "in {} but not installed.  Forgot to 'yarn install'?"
+                                    .format(
+                                        npm_name,
+                                        'dependency' if is_dep else 'devDependency',
+                                        manifest_file
+                                    )
+                                )
+                                self.notify_failure()
+                                raise linter.PermanentError()
+
+                        # Definitely not a Yarn PnP project
```

Comment:
Then here comes the inner loop searching upwards for binaries. If it fails we tell the user to run npm or yarn install.

Reply:
This must be exhaustive bc if we found a package.json we def break the outer loop!